text/0000-url-primitives.md

```diff
@@ -0,0 +1,352 @@
+
+All examples will have the following in common:
+```ts
+// router.js
+import EmberRouter, { URLManager } from '@ember/routing/router';
+import config from 'app-name/config/environment';
+
+// ... Example CustomURLManager here ...
+
+export default class Router extends EmberRouter {
+  location = config.locationType;
+  rootURL = config.rootURL;
+  urlManager = CustomURLManager; // Defined in Examples
+});
+
+Router.map(function() {
+  this.route('blogs', function() {
+    // index route would be the "list" of blogs
+    this.route('blog', { path: ':id' }, function() {
+      this.route('posts', function() {
+        // index route would be the "list" of posts
+        this.route('post', { path: ':id' });
+      })
+    });
+  })
+});
+```
+
+
+### Naïve Query Params without Controllers
+
+This example proposes the possibility of an alternative strategy for managing query params without the need for controllers. 
+Note that this would not change or alter the existing query param behavior in any way.
+
+```ts
+// app/router.js/ts
+import { inject as service } from '@ember/service';
+import qs from 'qs';
+
+class CustomURLManager extends URLManager {
+  @service router;
+
+  fromURL(url: string): RouteInfo { // /blogs/1/posts/2?foo=bar
+    let [path, query] = url.split('?');
+
+    let routeInfo = this.router.recognize(path);
+    // Because query params have no standardized way of (de)serialization,
+    // there has been no way to transform deep objects or arrays.
+    // This gives control over this process, allowing existing code that hacked
+    // around this limitation to be deleted.
+    let queryParams = qs.parse(query);
+
+    return {
+      ...routeInfo,
+      queryParams,
+    };
+  }
+
+  toURL(routeInfo: RouteInfo) {
+    let { 
+      mapInfo: {
+        segments // [blogs, :blogId, posts, :postId]
+      },
+      queryParams, // { foo: 'bar' }
+      dynamicSegments,
+    } = routeInfo;
+
+    let url = segments.map(segment => dynamicSegments[segment] || segment).join('/');
+
+    let query = qs.stringify(queryParams);
+
+    return `/${url}?${query}`; // => /blogs/1/posts/2?foo=bar
+  }
+}
+```
+
+### i18n routes
+
+In this scenario, we may want to internationalized route segments.
+
+For example, in English, we may want a route to be `/blogs/1/posts/2`, 
+but in Korean, `/블로그/1/게시물/2`.
+
+```ts
+// app/router.js/ts
+class CustomURLManager extends URLManager {
+  // current language: Korean
+  @service i18n;
+  @service router;
+
+  fromURL(url: string): RouteInfo { // /블로그/1/게시물/2?foo=bar
+    let [path, query] = url.split('?');
+    let segments = path.split('/');
+
+    // this dosen't have to be english, but it does have to match the names as
+    // defined in Router.map(...);
+    let english = 
+      path.split('/')
+          .segments.map(segment => {
+            return this.i18n.lookup(`routes.from.${segment}`, 'en-us') || segment;
```

Comment:
Wouldn't this imply that the `en-us` locale contains the mappings from all other languages' segments to the names used in the code? Shouldn't this mapping actually be part of every individual locale?

Reply:
> Wouldn't this imply that the en-us locale contains the mappings from all other languages' segments to the names used in the code?

yes.

> Shouldn't this mapping actually be part of every individual locale?

maybe. I don't know how molt people would want to implement this. Ember's route's are typically named in english, hence the approach here.

Reply:
I just wanted to note, that if used like this, it'd probably be wiser to remove the `en-us` in favor of the current locale.